lib/cucumber/formatter/json.rb

```diff
@@ -190,6 +190,7 @@ def create_data_table_value(data_table)
       def add_match_and_result(test_step, result)
         @step_or_hook_hash[:match] = create_match_hash(test_step, result)
         @step_or_hook_hash[:result] = create_result_hash(result)
+        result.embeddings.each { |e| embed(e['src'], e['mime_type'], e['label']) } if result.respond_to?(:embeddings)
```

Comment:
This is insanely nested and I may be barking up the wrong tree here. But do we care about the return value of this method.

Would there be any harm / change in code if we moved this method to in between the other 2.

I know this one looks like it's mutating each result to have new data attached to it (So the return in theory doesn't matter right?)

Reply:
Agreed, this formatter might benefit from being a visitor for results.

Swapping the lines should be possible, but I don't think we need to: `add_match_and_result()` is only called once and its return value is ignored.

Also, I guess any issue should have been catched by the tests, the json formatter looks well covered.

Reply:
So I've re-looked at this and have 2 new points (Again might be redundant).

From the method signature we are adding these result cases to the iVar `@step_or_hook_hash`

Now when we attach the result into 191 / 192, it is the result that is passed in. However here you are mutating the result (Altering an embeddings property).

I think it would be more prudent to move it up to 191 mutate it, then pass this new object into wherever it goes.

2nd point, can we alter the temp value of the block. Is e the embedding or the result.

I appreciate doing `embedding.embed` sounds a bit bad, but I think with something quite deeply nested having more information is better than less information

Reply:
I'm not sure to get your 1st point. Are you suggesting to write something like:
`@step_or_hook_hash[:embeddings] = create_embeddings(result)`

Reply:
Point 1 is you're attaching metadata to something, but after it's been attached somewhere else.

I think you should re-order it so it attaches metadata, then attaches it somewhere else.

Reply:
`embed()` pushes embeddings to `@step_or_hook_hash[:embeddings]`, it's not mutating `@step_or_hook_hash[:result]`.
I guess those two structures are required, because when testing Ruby code (no Wire protocol), when a step calls `embed()`, it is directly forwarded to the formatters' `embed()` method, filling
`@step_or_hook_hash[:embeddings]` before the call to `add_match_and_result()` (which will occur when the step ends).